java/src/main/native/ai_onnxruntime_OrtSession_SessionOptions.c

```diff
@@ -350,6 +350,85 @@ JNIEXPORT void JNICALL Java_ai_onnxruntime_OrtSession_00024SessionOptions_addFre
   (*jniEnv)->ReleaseStringUTFChars(jniEnv,dimensionName,cName);
 }
 
+/*
+ * Class:     ai_onnxruntime_OrtSession_SessionOptions
+ * Method:    addExternalInitializers
+ * Signature: (JJ[Ljava/lang/String;[J)V
+ */
+JNIEXPORT void JNICALL Java_ai_onnxruntime_OrtSession_00024SessionOptions_addExternalInitializers
+    (JNIEnv * jniEnv, jobject jobj, jlong apiHandle, jlong optionsHandle, jobjectArray namesArray, jlongArray handlesArray) {
+  (void) jobj; // Required JNI parameter not needed by functions which don't need to access their host object.
+  const OrtApi* api = (const OrtApi*)apiHandle;
+  OrtSessionOptions* options = (OrtSessionOptions*) optionsHandle;
+
+  size_t arrLength = (*jniEnv)->GetArrayLength(jniEnv, handlesArray);
+
+  const char** names = allocarray(arrLength, sizeof(char*));
+  if (names == NULL) {
+    // Nothing to cleanup, return and throw exception
+    throwOrtException(jniEnv, 1, "Not enough memory");
+    return;
+  }
+  jobject* javaNameStrings = allocarray(arrLength, sizeof(jobject));
+  if (javaNameStrings == NULL) {
+    goto cleanup_names;
+  }
+  const OrtValue** initializers = allocarray(arrLength, sizeof(OrtValue*));
+  if (initializers == NULL) {
+    goto cleanup_java_input_strings;
+  }
+
+  // Extract a C array of longs which are pointers to the input tensors.
+  // The Java-side objects store native pointers as 64-bit longs, and on 32-bit systems
+  // we cannot cast the long array to a pointer array as they are different sizes,
+  // so we copy the longs applying the appropriate cast.
+  jlong* initializersArr = (*jniEnv)->GetLongArrayElements(jniEnv, handlesArray, NULL);
+
+  for (size_t i = 0; i < arrLength; i++) {
+    // Extract the string chars and cast the tensor
+    javaNameStrings[i] = (*jniEnv)->GetObjectArrayElement(jniEnv, namesArray, (jint) i);
+    names[i] = (*jniEnv)->GetStringUTFChars(jniEnv, javaNameStrings[i], NULL);
+    initializers[i] = (const OrtValue*) initializersArr[i];
+  }
+
+  checkOrtStatus(jniEnv,api,api->AddExternalInitializers(options,names,initializers,arrLength));
```

Comment:
This function throws on errors? Won't that skip over the cleanup below?

Reply:
It throws an error up into the JVM, but the rest of the native function completes before control is returned to the Java code which propagates the exception. So the cleanup happens before any more Java code executes.

Reply:
Ah, ok. I was wondering how C was throwing exceptions, that makes sense. It's a Java thing. All good.